### PR TITLE
Refine View interfaces:

### PR DESCRIPTION
**Pull request by vinniefalco**
This shores up the View interface support for contextual
transaction processing by putting params in the View, and
provides support for replacing the open ledger with the
open MetaView.

Transaction metadata is now part of the View interface.
Stacked MetaViews correctly apply their transaction
metadata to the parent.

* Add lastCloseTime to View
* Add insertTx to View, implement in MetaView
* Add View::txExists for transaction checking
* Add Fees to View, cache fees in Ledger and MetaView
* Use ViewFlags in View
* Use tapENABLE_TESTING flag for features
* Use cached Fees in View
* Rename to ViewFlags
* Move FreezeHandling to View.h, remove ViewAPIBasics.h
* Remove BasicView::parent hack
* Remove calls to getLedger in Transactors

### CODE REVIEW

**Comment by ximinez** [comment]
Reviewers: @ximinez @nbougalis

**Comment by ximinez** [comment]
:+1: as of https://github.com/vinniefalco/rippled/commit/bc04666294c116e02cf184b1f358424defe4372e, though I know more changes are coming.

**Comment by vinniefalco** [comment]
Main commit message needs updating

**Comment by ximinez** [comment]
:+1: as of https://github.com/vinniefalco/rippled/commit/e8b12bb0919561ee4eebde81a164a069f6767346

**Comment by vinniefalco** [comment]
I'm not making any more changes, is this ready for a squash?

**Comment by ximinez** [comment]
Sure!

File: src/ripple/app/ledger/impl/MetaView.cpp [hunk 1]
```diff
@@ -42,34 +42,34 @@ namespace ripple {
...
 
-MetaView::MetaView (BasicView& parent,
-        bool openLedger)
-    : parent_ (&parent)
-    , mParams (openLedger
-        ? tapOPEN_LEDGER : tapNONE)
+MetaView::MetaView (BasicView const* parent,
+    ViewFlags flags, boost::optional<
+        uint256> const& key)
+    : parent_ (parent)
```
Comment (nbougalis) [inline src/ripple/app/ledger/impl/MetaView.cpp]: Consider constructor delegation:

    MetaView::MetaView (
        BasicView const* parent,
        ViewFlags flags,
        boost::optional<uint256> const& key)
            : MetaView (parent, parent->seq(), parent->time(), flags)
    {
        if (key)
            meta_.init (*key, seq_);
    }

File: src/ripple/app/ledger/impl/MetaView.cpp [hunk 2]
```diff
@@ -42,34 +42,34 @@ namespace ripple {
...
-    TransactionEngineParams tep)
-    : parent_(&*ledger)
-    , mParams (tep)
-{
-}
-
-MetaView::MetaView (MetaView& parent)
-    : parent_(&parent)
-    , mParams(parent.mParams)
+MetaView::MetaView (View const* parent)
```
Comment (nbougalis) [inline src/ripple/app/ledger/impl/MetaView.cpp]: Here too.

File: src/ripple/app/ledger/impl/MetaView.cpp (hunk 1 above)
Comment (vinniefalco) [inline src/ripple/app/ledger/impl/MetaView.cpp]: Changed in subsequent commits

**Comment by nbougalis** [comment]
:+1:

**Comment by vinniefalco** [comment]
Squashed

**Pull request closed without merging**